libmamba/src/core/progress_bar.cpp

```diff
@@ -342,206 +441,1521 @@ namespace mamba
+            f->set_format("{:>{}}").activate().set_width(0);
+        prefix.set_format("{:<{}}");
 
-    void HiddenProgressBar::print()
-    {
+        return *this;
     }
 
-    void HiddenProgressBar::set_full()
+    void ProgressBarRepr::compute_progress_value()
     {
-        if (!m_start_time_saved)
-        {
-            set_start();
-        }
-        p_manager->update_download_bar(m_total - m_current);
-    }
+        std::stringstream sstream;
+        auto width = progress.width(false);
 
-    void HiddenProgressBar::set_progress(size_t current, size_t /*total*/)
-    {
-        if (!m_start_time_saved)
+        if (!p_progress_bar->is_spinner())
         {
-            set_start();
+            if (width < 12)
+                sstream << std::ceil(p_progress_bar->progress()) << "%";
+            else
+            {
+                ProgressScaleWriter w{ width, "━", "╸", "━" };
+                double in_progress = static_cast<double>(p_progress_bar->in_progress())
+                                     / static_cast<double>(p_progress_bar->total()) * 100.;
+                sstream << w.repr(p_progress_bar->progress(), in_progress);
+            }
         }
-        size_t old_current = m_current;
-        m_current = current;
-        p_manager->update_download_bar(m_current - old_current);
-    }
-
-    void HiddenProgressBar::set_extracted()
+        else
+        {
+            if (width < 12)
+            {
+                std::vector<std::string> spinner
+                    = { "⣾", "⣽", "⣻", "⢿", "⣿", "⡿", "⣟", "⣯", "⣷", "⣿" };
+                auto pos
+                    = (static_cast<std::size_t>(std::round(p_progress_bar->progress())) % 50) / 5;
+                sstream << fmt::format("{:^4}", spinner[pos]);
+            }
+            else
+            {
+                auto pos = static_cast<std::size_t>(
+                    std::round(p_progress_bar->progress() * (width - 1) / 100.0));
+
+                std::size_t current_pos = 0, in_progress_pos = 0;
+
+                if (p_progress_bar->total())
+                {
+                    current_pos = static_cast<std::size_t>(
+                        std::floor(static_cast<double>(p_progress_bar->current())
+                                   / static_cast<double>(p_progress_bar->total()) * width));
+                    in_progress_pos = static_cast<std::size_t>(
+                        std::ceil(static_cast<double>(p_progress_bar->current()
+                                                      + p_progress_bar->in_progress())
+                                  / static_cast<double>(p_progress_bar->total()) * width));
+                }
+
+                auto spinner_half_width = width / 8;
+                for (std::size_t i = 0; i < width; ++i)
+                {
+                    if ((i < current_pos - 1 && current_pos > 0) || (current_pos == width))
+                    {
+                        sstream << fmt::format(
```

Comment:
I don't like that we're calling `fmt::format` inside a for loop to format single characters.
It would be enough to set the color once and then print all characters at once, no?

Reply:
you basically don't know if the next char will be blue, magenta or grey. 
there is probably room for improvement, for example with groups of same colors.
I can at least compute each character once and reuse it

Reply:
I think the idea was to use the fmt fill character stuff here: 

```c++
int count = 10;
fmt::format("{:━{}}", "", count);
```

but we'd need to compute `count` before hand which I'd hope is possible?